### PR TITLE
Docker based development for Jekyll

### PR DESCRIPTION
**Pull request by epugh**
### Description
Frustrated with Ruby install issues?  Using Docker already?   This let's you contribute documentation improvements without installing Ruby locally.

### Issues Resolved
_List any issues this PR will resolve, e.g. Closes [...]._

### Checklist
- [ x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Naarcha-AWS** [comment]
Love this @epugh! Let me know when it's ready for review.

**Comment by dtaivpp** [comment]
Ping

**Comment by epugh** [comment]
Hi all..   So I paired on this with @dtaivpp  and I *think* I have something working.  One potentially controversial decision was to lock the version of Ruby to `3.2.4`.  I polled my local Ruby group on their opinions on if the `Gemfile.lock` should be checked in and the consensus was that yes, it should be, and yes, you should pick a Ruby and stick with it.  By doing this, we don't regenerate the Gemfile.lock constantly, and we can upgrade dependencies more intentionally.   I would LOVE to hear if this works properly on a Linux and Window's environments....

**Comment by dtaivpp** [comment]
@epugh was the --net=host the solution for your use case? You think we should document that pattern?

**Comment by epugh** [comment]
> @epugh was the --net=host the solution for your use case? You think we should document that pattern?

I actually didn't a get chance to try that one.....

**Review by dtaivpp — CHANGES_REQUESTED** [review]
Works on Linux but I had a few comments to make it more usable.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+COPY Gemfile Gemfile.lock ./
+
+RUN bundle install
+
+# Expose the default Jekyll port
+EXPOSE 4000
+
+# Enable the link checker
+ENV JEKYLL_LINK_CHECKER="internal"
+
```
[the remark this answers is not in the document]
Comment (epugh) [reply Dockerfile:14]: Probably!   I haven't actually used ENV much in the past, so it felt "new" to me... !  Please make the change!

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+CMD ["bundle", "exec", "jekyll", "serve", "--host", "0.0.0.0", "--port", "4000", "--incremental", "--livereload", "--trace"]
```
[the remark this answers is not in the document]
Comment (epugh) [reply Dockerfile:15]: I love it!

[the remark this answers is not in the document]
Comment (epugh) [reply Dockerfile:15]: should we add a 1 liner description in the CONTRIBUTING.md file ?   "To run this on your droplet..."???

[the remark this answers is not in the document]
Comment (dtaivpp) [reply Dockerfile:15]: 100% I'll add that along with a bit more about the pattern.

**Review by dtaivpp — APPROVED** [review]
@kolchfa-aws I reviewed these and confirmed the steps outlined work and are documented correctly!

**Comment by Naarcha-AWS** [comment]
Test on Windows was successful.

**Comment by Naarcha-AWS** [comment]
@epugh: Once you take this out of draft mode, I'll give the language and formatting a proper review. Ping me when you're ready.

**Comment by epugh** [comment]
@Naarcha-AWS  it's no longer draft!   THanks for testing on windows!

**Comment by kolchfa-aws** [comment]
Thanks for validating the steps, @dtaivpp! Much appreciated. Let's get a review from @AMoo-Miki too (specifically fixing the Ruby version).

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -70,6 +70,23 @@ If you want to make minor changes to an existing file, you can use this approach
 
 If you're adding a new page or making major changes to the documentation, such as adding new images, sections, or styling, we recommend that you work in a local copy of the repository and test the rendered HTML before submitting a PR. 
 
+#### Setting up your Docker based copy of the repository
```
Comment (Naarcha-AWS) [inline CONTRIBUTING.md:73]:
```suggestion
#### Setting up a Docker-based copy of the repository

```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -70,6 +70,23 @@ If you want to make minor changes to an existing file, you can use this approach
...
+This assumes you have Docker installed.
```
Comment (Naarcha-AWS) [inline CONTRIBUTING.md:74]:
```suggestion
If you have Docker installed, you can use Docker to test any major changes to the documentation website locally by using the following steps:
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -70,6 +70,23 @@ If you want to make minor changes to an existing file, you can use this approach
...
+
+1. [Fork this repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo) and clone your fork.
```
Comment (Naarcha-AWS) [inline CONTRIBUTING.md:76]:
```suggestion
1. [Fork this repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo) and clone your fork in the directory of your choice.
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -70,6 +70,23 @@ If you want to make minor changes to an existing file, you can use this approach
...
+
+1. Navigate to your cloned repository.
+
+1. Build the documentation website Jekyll image via: `docker build -t documentation-website .`
```
Comment (Naarcha-AWS) [inline CONTRIBUTING.md:80]:
```suggestion
1. Build the documentation website Jekyll image by using the following command: 

    ```
    docker build -t documentation-website .
    ```
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -70,6 +70,23 @@ If you want to make minor changes to an existing file, you can use this approach
...
+
+1. Start Jekyll via: `docker run -p 4000:4000 -v $(pwd):/app documentation-website`
```
Comment (Naarcha-AWS) [inline CONTRIBUTING.md:82]:
```suggestion
1. Start Jekyll from inside the image using the following command: 

   ```
   docker run -p 4000:4000 -v $(pwd):/app documentation-website
   ```
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -70,6 +70,23 @@ If you want to make minor changes to an existing file, you can use this approach
...
+
+Be aware that it takes a while to build the site the first time, eventually you will see `Server running... press ctrl-c to stop`.
```
Comment (Naarcha-AWS) [inline CONTRIBUTING.md:84]:
```suggestion
When building the site for the first time, it'll take some time for the site to build. When your Jekyll instance is ready, you will see `Server running... press ctrl-c to stop`.
```

**Review by Naarcha-AWS — CHANGES_REQUESTED** [review]
This changes suggested make the formatting more in line with the next section.

**Comment by Naarcha-AWS** [comment]
@epugh: Per the advice of @AMoo-Miki, the `gemfile` shouldn't prevent someone from using a different version of ruby if they so choose. One alternative is to make a gemfile specific for Docker.

**Comment by hdhalter** [comment]
@dtaivpp , @epugh - Is Miki's comment a blocker for pushing this up?

**Comment by dtaivpp** [comment]
@hdhalter I think that comes down to the preferences of the project. It's pretty standard to standardize on a single version of ruby. That way changes that are made in one persons environment should be guaranteed to work in production with the same ruby version. 

@AMoo-Miki I'm curious what advantage there is to allowing users to use any ruby version they'd like. That feels like a risk

**Comment by epugh** [comment]
Just went to add some docs and ran `build.sh` and hit the 

```
Could not find jekyll-last-modified-at-1.3.0, i18n-1.14.3, rouge-4.2.0, posix-spawn-0.3.15, public_suffix-5.0.4, concurrent-ruby-1.2.3, sass-embedded-1.71.1-x86_64-darwin, ffi-1.16.3, google-protobuf-3.25.3-x86_64-darwin, rb-inotify-0.10.1 in locally installed gems
```
message.  

The big change that impacts non Docker users in this set up is that we check into the repo `Gemfile.lock` and specify the version of Ruby to use, instead of leaving it open....    So you have to use a specific Ruby...  Though we will then know all the versions of the gems you use ;-).   This is my preference.

If that is too much, I could look at having Docker install ruby and the gems at start up time.  That would make firing up the docker image a lot slower, but might be a more gentle introduction of Docker into the tool chain....    Thoughts @AMoo-Miki ?

**Comment by epugh** [comment]
This was solved in #8220!  Thanks @hainenber for getting this problem solved!

**Pull request closed without merging**